dwave/samplers/sa/sampler.py

```
@@ -379,6 +386,13 @@ def sample(self, bqm: dimod.BinaryQuadraticModel,
 
         response.change_vartype(original_vartype, inplace=True)
 
+        # Update timing info last to capture the full postprocessing time
+        response.info.update(dict(timing=dict(
+            preprocessing=t1 - t0,
+            sampling=t2 - t1,
+            postprocessing=datetime.now() - t2
+        )))
```

Two obvious issues I see here are:
1. `info` is now not JSON-serializable anymore, hence `SampleSet.to_serializable()` will fail
2. it's confusing to use `.timing` both here and for the QPU sampler timing info -- which is a very different beast (for better or worse)

To resolve these issues, I propose:
1. add support for `timedelta` serialization to dimod (could be serialized as [ISO 8601 duration](https://en.wikipedia.org/wiki/ISO_8601#Durations), e.g. `PT0.000123S` to represent 123 us).
2. use a different dict key here. I don't have a great suggestion, but anything different from `timing` will work. E.g. `runtime`, `wallclock`, `tictoc`, `runstats`...

---

For (2), we already using `timing` in both the hybrid solvers and the QPU with different contents and that has not caused any issues that I know of. So IMO it makes sense to continue that convention.

Similarly for (1), the hybrid solvers already return their timings as simple floats representing seconds. Rather than introduce more ad hoc serialization rules why not just use floats and document them? I understand the desire for units, but as far as I know the hybrid solver timing info has never been confusing for users.

---

RE (1):
I favoured `timedelta` because it contains unit information and is easy to manipulate. However, if serialization is needed, this would introduce unnecessarily complex workflows especially for benchmarking. See, for example, [parsing ISO8601 duration to timedelta object](https://stackoverflow.com/questions/36976138/is-there-an-easy-way-to-convert-iso-8601-duration-to-timedelta).

I prefer storing timings as native floats. In addition to documenting the units, I suggest using dictionary keys that include units, e.g., `timing_us` or `sampling_seconds`, for two reasons:
1. No ambiguity in naming/returned units.
2. If the units of returned durations eventually change, the keys would need to reflect this change. I view this as a pro. Example: an existing pipeline uses `time_us` and the timing is updated to `timing_seconds`, the pipeline will explicitly throw an error instead of fail silently under wrong assumptions.


(No comment on (2))

---

I am fine with floats with units in names as a reasonable compromise. @randomir , thoughts?

---

Re: serialization -- if `timedelta` seems infeasible atm, I guess using float seconds (with double precision) is sufficient to accurately represent up to a microsecond. Unlikely we'll need a higher resolution than than in Ocean.

---

@randomir , you're totally right. I had a brain fart. We don't include a timing nested dict.

With that in mind, perhaps we can follow that precedent here and just use rather than including a nested structure
```
info = {"preprocessing_time_s": ...}
```

---

+1 to @randomir 's comment: the timing for local classical solvers (e.g., SA) is more comparable to the QPU's. With that, I think nested timing is more consistent with the timing info in DWaveSampler (`print(DWaveSamplers().sample_ising({}, {}).info['timing']`).

---

Ok, to summarize.

QPU samplers return `info["timing"] = {"qpu_sampling_time": ..., "qpu_readout_time_per_sample": ...}`, with values in integer microseconds.

HSS samplers return `info = {"qpu_access_time": ..., "charge_time": ..., "run_time": ...}`, with values in integer microseconds.

And now here for local classical samplers we have a few options. I think we should pick something that's different enough so it can be easily differentiated going forward by an (external) uninitiated "processor" (think a function unaware of the sampler type used).

For example: `info["sampler_timing"] = {"preprocessing_time": ..., "sampling_time": ..., "postprocessing_time": ...}`, with values in integer microseconds.

---

A "processor" that is looking for a specific structure will fail just as easily if the keys within the `"timing"` nested dict are missing as if we changed the name of the outer dict. In either case the user will get an error if they are assuming it will be present. `"sampler_timing"` just seems redundant and ugly to me, without providing any additional clarity to the user. Being consistent about putting timing info into a `"timing"` inner dict is at least the beginning of a useful convention.

---

Both nested and flat dictionaries make sense to me.
We should include units in keys though, e.g.,
`info["sampler_timing"] = {"preprocessing_time_us": ..., "sampling_time_ms": ..., "postprocessing_time_s": ...}`, or
`info["sampler_timing_us"] = {"preprocessing_time": ...}`.

---

`"sampler_timing"` might be ugly, and I don't care about the actual string, but the idea was that *that is the beginning* of a useful convention.

And my point about a "processor" was that whatever we do here, any code that processes timing will have to support *yet another convention*, so why not make this *the convention* going forward.

But sure, we can conflate it with `"timing"`.